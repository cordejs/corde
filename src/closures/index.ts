export * from "./group";
export * from "./test";
