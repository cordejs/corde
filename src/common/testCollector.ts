import { Queue } from "../data-structures";
import { IAssertionProps, IGroup, ITest, TestFunctionType, VoidLikeFunction } from "../types";

/**
 * Contain all information of data collected from files in runtime test
 * collection.
 * @internal
 */
class TestCollector {
  private constructor() {
    this.isInsideGroupClausure = false;
    this.isInsideTestClausure = false;
    this.groups = [];
    this.beforeEachFunctions = new Queue();
    this.afterAllFunctions = new Queue();
    this.beforeStartFunctions = new Queue();
    this.afterEachFunctions = new Queue();

    this.testClousureFunction = new Queue();
    this.groupClousureFunction = new Queue();

    this.tests = [];
    this.assertions = [];
    this.isolatedFunctions = [];
    this.testsFunctions = [];
    this.testsPass = 0;
    this.testsFailed = 0;
  }

  private static _instance: TestCollector;
  isInsideGroupClausure: boolean;
  isInsideTestClausure: boolean;

  assertions: IAssertionProps[];
  tests: ITest[];
  groups: IGroup[];

  beforeStartFunctions: Queue<VoidLikeFunction>;
  afterAllFunctions: Queue<VoidLikeFunction>;
  beforeEachFunctions: Queue<VoidLikeFunction>;

  afterEachFunctions: Queue<VoidLikeFunction>;
  testsFunctions: TestFunctionType[];
  isolatedFunctions: TestFunctionType[];

  testsPass: number;
  testsFailed: number;

  private testClousureFunction: Queue<VoidLikeFunction>;
  private groupClousureFunction: Queue<VoidLikeFunction>;

  static getInstance() {
    if (!TestCollector._instance) {
      TestCollector._instance = new TestCollector();
    }
    return TestCollector._instance;
  }

  addTestFunction(testFunction: TestFunctionType) {
    if (testFunction) {
      if (this.isInsideGroupClausure || this.isInsideTestClausure) {
        this.testsFunctions.push(testFunction);
      } else {
        this.isolatedFunctions.push(testFunction);
      }
    }
  }

  isInsideTestClausureFunctions() {
    return this.testsFunctions && this.testsFunctions.length > 0;
  }

  hasIsolatedTestFunctions() {
    return this.isolatedFunctions && this.isolatedFunctions.length > 0;
  }

  cloneTestFunctions() {
    return this.testsFunctions.map((m) => m);
  }

  cloneIsolatedTestFunctions() {
    return this.isolatedFunctions.map((m) => m);
  }

  clearTestFunctions() {
    this.testsFunctions = [];
  }

  clearIsolatedTestFunctions() {
    this.isolatedFunctions = [];
  }

  cleanAll() {
    this.tests = [];
    this.testsFunctions = [];
    this.groups = [];
  }

  addToGroupClousure(fn: () => void | Promise<void>) {
    this.groupClousureFunction.enqueue(fn);
  }

  async executeGroupClojure() {
    return await this.groupClousureFunction?.executeWithCatchCollectAsync();
  }

  addToTestClousure(fn: () => void | Promise<void>) {
    this.testClousureFunction.enqueue(fn);
  }

  async executeTestClojure() {
    return await this.testClousureFunction?.executeWithCatchCollectAsync();
  }
}

const testCollector = TestCollector.getInstance();
export { testCollector };
