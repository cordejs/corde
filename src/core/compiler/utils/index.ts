export * from "./import-js-in-directory";
export * from "./was-file-modified";
