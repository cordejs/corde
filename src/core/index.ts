export * from "./Group";
export * from "./TestFile";
export * from "./Config";
export * from "./debug";
export * from "./Events";
export * from "./injectGlobals";
export * from "./internalEvents";
export * from "./printHookError";
export * from "./Reader";
export * from "./runtime";
export * from "./summary";
export * from "./testCollector";
export * from "./TestExecutor";
export * from "./tsRegister";
export * from "./CordeBot";
