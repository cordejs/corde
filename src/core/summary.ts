/* eslint-disable no-console */
import { TEXT_EMPTY, TEXT_FAIL, TEXT_PASS } from "../const";
import { IRunnerReport } from "../types";

interface ISummaryLine {
  label: string;
  total: number;
  fail: number;
  success: number;
  totalEmpty: number;
}

class Summary {
  print(runnerReport: IRunnerReport) {
    let message = "\n";
    message += this.buildTestFilesSummary(runnerReport) + "\n";
    message += this.buildTestsSummary(runnerReport) + "\n";
    if (runnerReport.testTimer) {
      message += "Time:       " + runnerReport.testTimer;
    }
    console.log(message);
    return message;
  }

  private buildTestFilesSummary(runnerReport: IRunnerReport) {
    return this.buildSummaryLine({
      fail: runnerReport.totalTestFilesFailed,
      label: "Test Files: ",
      success: runnerReport.totalTestFilesPassed,
      totalEmpty: runnerReport.totalEmptyTestFiles,
      total: runnerReport.totalTestFiles,
    });
  }

  private buildTestsSummary(runnerReport: IRunnerReport) {
    return this.buildSummaryLine({
      fail: runnerReport.totalTestsFailed,
      label: "Tests:      ",
      success: runnerReport.totalTestsPassed,
      totalEmpty: runnerReport.totalEmptyTests,
      total: runnerReport.totalTests,
    });
  }

  private buildSummaryLine(data: ISummaryLine) {
    const { label, fail, success, total, totalEmpty } = data;

    const partialLine: string[] = [];
    if (success > 0) {
      partialLine.push(TEXT_PASS(`${success} passed`));
    }

    if (fail > 0) {
      partialLine.push(TEXT_FAIL(`${fail} failed`));
    }

    if (totalEmpty > 0) {
      partialLine.push(TEXT_EMPTY(`${totalEmpty} empty`));
    }

    if (partialLine.length) {
      return `${label}${this.buildSummaryLinePhrase(partialLine)}. ${total} total`;
    }

    if (total === 0) {
      return `${label}${TEXT_EMPTY(`${total} total`)}`;
    }
    return `${label}${total} total`;
  }

  /**
   * Builds the summary line phrase
   *
   * @example
   *
   * ["1 passed", "2 fail"] => '1 passed and 2 fail'
   * ["1 passed", "2 fail", "1 empty"] => '1 passed, 2 fail and 1 empty'
   */
  private buildSummaryLinePhrase(parts: string[]) {
    let builtLine = "";

    for (let i = 0; i < parts.length; i++) {
      if (parts.length > 2 && i < 2) {
        builtLine += parts[i] + ", ";
      } else {
        builtLine += parts[i] + " and ";
      }
    }

    // removes the last 'and '
    return builtLine.substring(0, builtLine.length - 5);
  }
}

const summary = new Summary();
export { summary };
