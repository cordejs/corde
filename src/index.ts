export { group, test, group as describe, test as it } from "./closures";
export { expect } from "./expect";
export { afterAll, beforeStart, afterEach, beforeEach } from "./hooks";
export * from "./types/general";

import { corde } from "./corde";
export const bot = corde.bot;

export default corde;
