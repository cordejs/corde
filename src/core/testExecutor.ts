import chalk from "chalk";
import { printHookErrors } from "../common/printHookError";
import {
  MESSAGE_TAB_SPACE,
  TAG_FAIL,
  TAG_PASS,
  TAG_PENDING,
  TEST_FAIL_ICON,
  TEST_PASSED_ICON,
  TEST_RUNNING_ICON,
  TEXT_EMPTY,
  TEXT_FAIL,
  TEXT_PASS,
} from "../consts";
import { Queue } from "../data-structures";
import {
  IRunnerReport,
  ISemiRunnerReport,
  ITest,
  ITestReport,
  Nullable,
  VoidLikeFunction,
} from "../types";
import { stringIsNullOrEmpty, Timer } from "../utils";
import { LogUpdate } from "../utils";
import { TestFile } from "../common/TestFile";
import { Group } from "../common/Group";
import { runtime } from "../common/runtime";
import { TestError } from "../errors";

type ReportStatusType = "pass" | "fail" | "empty";

/**
 * @internal
 */
export class TestExecutor {
  private _logUpdate: LogUpdate;
  private _semiReport!: ISemiRunnerReport;

  constructor(logUpdate: LogUpdate) {
    this._logUpdate = logUpdate;
    this.initReport();
  }

  private initReport() {
    this._semiReport = {
      totalTests: 0,
      totalTestFiles: 0,
      totalTestsPassed: 0,
      totalTestsFailed: 0,
      totalTestFilesPassed: 0,
      totalTestFilesFailed: 0,
      totalEmptyTestFiles: 0,
      totalEmptyTests: 0,
    };
  }

  async runTestsAndPrint(testFiles: TestFile[]): Promise<IRunnerReport> {
    this.initReport();
    const testsTimer = new Timer();
    testsTimer.start();

    for (const testFile of testFiles) {
      await this.executeTestFile(testFile);
    }

    const testsDiff = testsTimer.stop();

    return {
      testTimer: testsDiff[0],
      ...this._semiReport,
    };
  }

  private async executeTestFile(testFile: TestFile) {
    if (testFile.isEmpty()) {
      this._logUpdate.append(`${TAG_PENDING("EMPTY")}  ${testFile.path}`);
      this._logUpdate.persist();
      this._semiReport.totalEmptyTestFiles++;
      this._semiReport.totalTestFiles++;
      return;
    }

    const testFileTimer = new Timer();
    this._semiReport.totalTestFiles++;

    testFileTimer.start();
    const logIndex = this._logUpdate.append(`${TAG_PENDING()}  ${testFile.path}`);

    await this.executeHookFunction(testFile.beforeAllHooks);

    const status = await this.executeClosures(testFile.closures, testFile);

    await this.executeHookFunction(testFile.afterAllHooks);

    const _diff = testFileTimer.stop();

    let fileLabel = TAG_PASS();
    let fileNameLabel = testFile.path;

    if (status === "fail") {
      fileLabel = TAG_FAIL();
      fileNameLabel = chalk.red(fileNameLabel);
    }

    this._logUpdate.updateLine(
      logIndex,
      `${fileLabel}  ${fileNameLabel}   ${chalk.cyan(_diff[0])}`,
    );

    this._logUpdate.persist();

    if (status === "pass") {
      this._semiReport.totalTestFilesPassed++;
    } else if (status === "fail") {
      this._semiReport.totalTestFilesFailed++;
    } else {
      this._semiReport.totalEmptyTestFiles++;
    }
  }

  private async executeGroup(group: Group, testFile: TestFile) {
    let status: ReportStatusType = "pass";
    await this.executeHookFunction(group.beforeAllHooks);
    status = await this.executeClosures(group.closures, testFile);
    await this.executeHookFunction(group.afterAllHooks);
    return status;
  }

  private async executeClosures(closures: (ITest | Group)[], testFile: TestFile) {
    let status: ReportStatusType = "pass";
    for (const closure of closures) {
      if (closure instanceof Group) {
        status = (await this.executeGroup(closure, testFile)) as ReportStatusType;
      } else {
        const report = await this.executeTest(closure, testFile);

        if (!report) {
          status = "empty";
        } else {
          status = report.pass ? "pass" : "fail";
        }
      }
    }
    return status;
  }

  private async executeTest(test: ITest, testFile: TestFile, group?: Group) {
    const testTimer = new Timer();
    testTimer.start();
    let logPosition = 0;

    const testName = await test.toResolveName();
    const testText = this.createTestText(testName);
    logPosition = this._logUpdate.appendLine(testText);

    const report = await this.runTest(test, testFile, group);

    this._semiReport.totalTests++;
    if (!report) {
      this._semiReport.totalEmptyTests++;
    }

    const testDiff = testTimer.stop();

    const testNameLabel = this.testReportLabelFunction(report);

    const formatedGroupName = !stringIsNullOrEmpty(group?.name) ? group?.name + " -> " : "";

    if (stringIsNullOrEmpty(testName)) {
      this._logUpdate.updateLine(
        logPosition,
        `${MESSAGE_TAB_SPACE}${testNameLabel(
          formatedGroupName + "<empty test name>",
        )}   ${chalk.cyan(testDiff[0])}`,
      );
    } else {
      this._logUpdate.updateLine(
        logPosition,
        `${MESSAGE_TAB_SPACE}${testNameLabel(formatedGroupName + testName)}   ${chalk.cyan(
          testDiff[0],
        )}`,
      );
    }

    this.printReportAndUpdateRunnerReport(report);

    return report;
  }

  private testReportLabelFunction(report?: ITestReport) {
    if (!report) {
      return (text: string) => TEXT_EMPTY(" " + TEST_RUNNING_ICON + " " + text + " (empty)");
    }

    if (!report.pass) {
      return (text: string) => TEXT_FAIL(TEST_FAIL_ICON + " " + text);
    }
    return (text: string) => TEXT_PASS(TEST_PASSED_ICON + " " + text);
  }

  private printReportAndUpdateRunnerReport(report?: ITestReport) {
    if (!report) {
      return;
    }

    if (report.pass) {
      this._semiReport.totalTestsPassed++;
    } else {
      this._semiReport.totalTestsFailed++;
    }

    this.printReportData(report);
  }

  private printReportData(report: ITestReport) {
    if (report.message) {
      this._logUpdate.appendLine(report.message);
    }

    if (!report.pass && report.trace) {
      this._logUpdate.appendLine(report.trace + "\n");
    }
  }

  private createTestText(testName?: string | number | boolean) {
    if (stringIsNullOrEmpty(testName)) {
      return this.createTestTextByStatus("<empty test name>");
    }
    return this.createTestTextByStatus(testName);
  }

  private createTestTextByStatus(testName?: string | number | boolean) {
    const icon = TEST_RUNNING_ICON;
    return `${MESSAGE_TAB_SPACE}${icon} ${testName}`;
  }

  async runTest(test: ITest, testFile: TestFile, group?: Group) {
    let report: Nullable<ITestReport> = undefined;
    // before e after hooks will run just one time
    // for test structure (if the hook fail)
    let keepRunningBeforeEachFunctions = true;
    let keepRunningAfterEachFunctions = true;

    if (keepRunningBeforeEachFunctions) {
      const testFileHookOk = await this.executeHookFunction(testFile.beforeEachHooks);
      const groupHookOk = await this.executeHookFunction(group?.beforeEachHooks);
      keepRunningBeforeEachFunctions = testFileHookOk && groupHookOk;
    }

    const onTestEnd = (_report: ITestReport) => {
      report = _report;
    };

    runtime.internalEvents.on("test_end", onTestEnd);

    try {
      await test.action();
    } catch (error) {
      if (error instanceof TestError) {
        report = {
          message: error.message,
          pass: error.pass,
          testName: error.testName,
          trace: error.trace,
        };
      }
    }

    runtime.internalEvents.removeListener("test_end", onTestEnd);

    if (keepRunningAfterEachFunctions) {
      const testFileHookOk = await this.executeHookFunction(testFile.afterEachHooks);
      const groupHookOk = await this.executeHookFunction(group?.afterEachHooks);
      keepRunningAfterEachFunctions = testFileHookOk && groupHookOk;
    }

    return report;
  }

  private async executeHookFunction(queues?: Queue<VoidLikeFunction>) {
    if (!queues) {
      return true;
    }

    const _functionErrors = await queues.executeWithCatchCollectAsync();
    if (_functionErrors && _functionErrors.length) {
      printHookErrors(_functionErrors);
      return false;
    }
    return true;
  }
}
