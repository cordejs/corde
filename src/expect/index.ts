/* eslint-disable no-console */
import chalk from "chalk";
import { runtime } from "../common/runtime";
import { ITestProps, ITestReport } from "../types";
import { corde } from "../types/globals";
import { getStackTrace, typeOf } from "../utils";
import { any } from "./asymmetricMatcher";
import * as matchers from "./matchers";

interface IMatcher {
  (props: ITestProps, ...args: any[]): ITestReport;
}

type KeyOfMatcher = keyof typeof matchers;

function pickFn(name: KeyOfMatcher) {
  return matchers[name] as any as IMatcher;
}

function createMatcherFn(matcher: string, isNot: boolean, expected: any, isDebug: boolean) {
  return (...args: any[]): ITestReport | void => {
    // If someone pass expect.any, we must invoke it to return
    // the Any matcher.

    if (expected === any) {
      args = [expected(), ...args];
    } else {
      args = [expected, ...args];
    }

    args = args.map((arg) => {
      if (arg === any) {
        return arg();
      }
      return arg;
    });

    const trace = getStackTrace(undefined, true, matcher);
    try {
      const matcherFn = pickFn(matcher as KeyOfMatcher);

      const props: ITestProps = {
        isNot,
        expectedColorFn: chalk.bold,
        receivedColorFn: chalk.bold,
        createHint: (...paramsName: string[]) => {
          return `expect(${chalk.green("expected")}).${isNot ? "not." : ""}${matcher}(${paramsName
            .map((param) => chalk.red(param))
            .join(",")})`;
        },
        formatValue: (value: any) => {
          if (typeof value === "symbol") {
            return value.toString();
          }

          if (typeOf(value) === "object") {
            if (Object.keys(value).length === 0) {
              return "{}";
            }

            return "{ ... }";
          }

          if (
            typeof value === "string" &&
            (value.length === 0 || (!value.startsWith("'") && !value.endsWith("'")))
          ) {
            return `'${value}'`;
          }

          if (Array.isArray(value)) {
            if (value.length > 0) {
              return "[...]";
            }
            return "[]";
          }

          return value;
        },
      };

      const report = matcherFn.bind(props, ...args)();
      if (!report.pass) {
        report.trace = trace;
      }

      runtime.internalEvents.emit("test_end", report);

      if (isDebug) {
        return report;
      }
    } catch (error) {
      const failedReport: ITestReport = {
        pass: false,
        message: handleError(error),
        trace: trace,
      };
      runtime.internalEvents.emit("test_end", failedReport);
      if (isDebug) {
        return failedReport;
      }
    }
  };
}

function handleError(error: any) {
  if (error instanceof Error) {
    return error.message;
  }
  return error;
}

function createLocalExpect(isDebug: boolean) {
  let localExpect: any = {};

  localExpect = (expected: any) => {
    const _expect: any = {};
    _expect.not = {};
    Object.getOwnPropertyNames(matchers).forEach((matcher) => {
      _expect[matcher] = createMatcherFn(matcher, false, expected, isDebug);
      _expect.not[matcher] = createMatcherFn(matcher, true, expected, isDebug);
    });
    return _expect;
  };

  localExpect.any = any;

  return localExpect;
}

export const expect = createLocalExpect(false) as corde.IExpect;

type DebugExpectType<T> = {
  [P in keyof T]: T[P] extends (...args: any[]) => any
    ? (...params: Parameters<T[P]>) => { pass: boolean; message: string }
    : DebugExpectType<T[P]>;
};

export interface IDebugExpect {
  <T extends any>(value: T): DebugExpectType<corde.IMatchersWithNot<T>>;
  any(...classType: any[]): any;
}

/**
 * Testing object for corde
 * This is not used in production
 *
 * @internal
 */
export const cordeExpect = createLocalExpect(true) as IDebugExpect;
