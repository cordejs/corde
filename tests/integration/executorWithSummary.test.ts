import { group, test as _test } from "../../src/closures";
import { expect as _expect } from "../../src/expect";
import { summary } from "../../src/core/summary";
import { TestExecutor } from "../../src/core/testExecutor";
import { LogUpdate } from "../../src/utils";
import { mockTimer } from "../mocks/mockTimer";
import { removeANSIColorStyle } from "../testHelper";
import { testCollector } from "../../src/core/testCollector";

let logUpdate: LogUpdate;
let testRunner: TestExecutor;

mockTimer();

beforeEach(() => {
  logUpdate = new LogUpdate();
  testRunner = new TestExecutor(logUpdate);
  testCollector.clearTestFiles();
  testCollector.createTestFile("test file");
});

it("should print report for 1 test file, 1 test closure and 1 failed function and 1 passed function", async () => {
  group("group", () => {
    test("test closure", () => {
      _expect(1).toEqual(1);
      _expect(1).toEqual(1);
    });
  });

  await testCollector.executeGroupClojure();
  const report = await testRunner.runTestsAndPrint(testCollector.testFiles);
  const summaryStder = summary.print(report);
  expect(removeANSIColorStyle(logUpdate.stdout + summaryStder)).toMatchSnapshot();
});
