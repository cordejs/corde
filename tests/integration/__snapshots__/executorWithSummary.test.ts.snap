// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print report for 1 test file, 1 test closure and 1 failed function and 1 passed function 1`] = `
"
Test Files: 1 passed. 1 total
Tests:      1 passed. 1 total
Time:       10ms"
`;
